pkg/fuzzer/job.go

```diff
@@ -0,0 +1,390 @@
+	stableSignal signal.Signal, inputCover cover.Cover, rawCover []uint32,
+	stop bool) {
+	const signalRuns = 3
+	// TODO: run all signalRuns in parallel?
+	var notExecuted int
+	for i := 0; i < signalRuns; i++ {
+		result := fuzzer.exec(job, &Request{
+			Prog:         job.p,
+			NeedSignal:   true,
+			NeedCover:    true,
+			NeedRawCover: fuzzer.Config.FetchRawCover,
+			stat:         statTriage,
+		})
+		if result.Stop {
+			stop = true
+			return
+		}
+		info := result.Info
+		if !reexecutionSuccess(info, &job.info, job.call) {
+			// The call was not executed or failed.
+			notExecuted++
+			if notExecuted >= signalRuns/2+1 {
+				return // if happens too often, give up
+			}
+			continue
+		}
+		thisSignal, thisCover := getSignalAndCover(job.p, info, job.call)
+		if len(rawCover) == 0 && fuzzer.Config.FetchRawCover {
+			rawCover = append([]uint32{}, thisCover...)
+		}
+		if stableSignal.Empty() {
+			stableSignal = newSignal.Intersection(thisSignal)
+		} else {
+			stableSignal = stableSignal.Intersection(thisSignal)
+		}
+		if stableSignal.Empty() {
+			return
+		}
+		inputCover.Merge(thisCover)
+	}
+	return
+}
+
+func (job *triageJob) minimize(fuzzer *Fuzzer, newSignal signal.Signal) (stop bool) {
+	const minimizeAttempts = 3
+	job.p, job.call = prog.Minimize(job.p, job.call, false,
+		func(p1 *prog.Prog, call1 int) bool {
+			if stop {
+				return false
+			}
+			for i := 0; i < minimizeAttempts; i++ {
+				result := fuzzer.exec(job, &Request{
+					Prog:       p1,
+					NeedSignal: true,
+					stat:       statMinimize,
+				})
+				if result.Stop {
+					stop = true
+					return false
+				}
+				info := result.Info
+				if !reexecutionSuccess(info, &job.info, call1) {
+					// The call was not executed or failed.
+					continue
+				}
+				thisSignal, _ := getSignalAndCover(p1, info, call1)
+				if newSignal.Intersection(thisSignal).Len() == newSignal.Len() {
+					return true
+				}
+			}
+			return false
+		})
+	return stop
+}
+
+func reexecutionSuccess(info *ipc.ProgInfo, oldInfo *ipc.CallInfo, call int) bool {
+	if info == nil || len(info.Calls) == 0 {
+		return false
+	}
+	if call != -1 {
+		// Don't minimize calls from successful to unsuccessful.
+		// Successful calls are much more valuable.
+		if oldInfo.Errno == 0 && info.Calls[call].Errno != 0 {
+			return false
+		}
+		return len(info.Calls[call].Signal) != 0
+	}
+	return len(info.Extra.Signal) != 0
+}
+
+func getSignalAndCover(p *prog.Prog, info *ipc.ProgInfo, call int) (signal.Signal, []uint32) {
+	inf := &info.Extra
+	if call != -1 {
+		inf = &info.Calls[call]
+	}
+	return signal.FromRaw(inf.Signal, signalPrio(p, inf, call)), inf.Cover
+}
+
+type smashJob struct {
+	p    *prog.Prog
+	call int
+}
+
+func (job *smashJob) priority() priority {
+	return priority{smashPrio}
+}
+
+func (job *smashJob) run(fuzzer *Fuzzer) {
+	fuzzer.Logf(2, "smashing the following program (call=%d):\n%s",
+		job.call, string(job.p.Serialize()))
```

Comment:
Some comment re p.Serialize().
If we want to have expensive calls in Logf, we need to add a predicate function that checks if level if enabled.
Or otherwise avoid expensive calls (log p.String).
IIRC if fuzzer starts logging all full serialized programs that creates unconsumable amount of output. Have you read full programs during debugging?

Reply:
> Have you read full programs during debugging?

At the beginning, when things barely worked, yes, it was helpful. Now I think we don't need that level of detail indeed. 

I've simplified the log line.